Implement BigDecimal

vegai opened the pull request:
This PR adds a BigDecimal implementation. 

The motivation for having BigDecimals is to have an arbitrary precision exact decimal type. This is required at least in financial calculations, where the precision of Floats and is not enough and BigRationals may have other problems due to denominator potentially being something else than 10**x. At least Ruby, Python and Java contain something like this in their standard libs.

The design is mostly adapted from bigdecimal-rs written for Rust. It holds its actual value in a BigInt, and a scale (decimal point place) in UInt64. It currently contains basic + - * / arithmetics. Possible future needs: implementation of ** (power) and supporting scientific "123E45" notation.

Please critique liberally. This would be my first larger Crystal contribution.

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+  # Set maximum iterations used in division operation. This implicitly
+  # defines the maximum precision of divisions in cases where the
+  # division is not exact.
+  def self.set_max_div_iterations(x : Int)
```
Please use descriptive names for arguments.

vegai replied:
=> new_max_div

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+  # Create a new `BigDecimal` from a String.
+  #
+  # Allows only valid number strings with an optional negative sign.
+  def initialize(s : String)
```
ditto

MakeNowJust replied:
> Use `ditto` to use the same comment as in the previous declaration.

https://crystal-lang.org/docs/conventions/documenting_code.html

vegai replied:
=> str

@MakeNowJust I guess this comment didn't apply here?

Sija replied:
@MakeNowJust Quoted remark applies to documentation, not code-review comments, mind you.

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+  end
+
+  # from Int
+  def initialize(i : Int)
```
ditto

vegai replied:
=> num

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+  end
+
+  # from Float is not supported due to precision loss risks. This call fails at compile time.
+  def initialize(f : Float)
```
ditto

vegai replied:
=> num

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+  end
+
+  private def power_ten_to(x : Int) : Int
+    BigInt.new(10) ** x
```
```cr
TEN ** x
```

vegai replied:
=> TEN

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+end
+
+struct Float
+  # from Float is not supported due to precision loss risks. This call fails at compile time.
```
I'd use full sentence here, like: "Casting from `Float` is not supported due to ..."

vegai replied:
Marked this Float and all the other similar ones too.

vegai commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+end
+
+struct Float
+  # from Float is not supported due to precision loss risks. This call fails at compile time.
```
So changed.

bew replied:
@vegai you should use `` ` `` around `Float` in your comment: ``Casting from `Float` ...``

akzhan commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+  include Comparable(BigDecimal)
+
+  # Convert `Int` to `BigDecimal`
+  def to_big_d
```
Looks like you forget to declare

```crystal
struct BigDecimal
  def to_big_d
    self
  end
end
```

vegai replied:
Yes. Added.

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+    elsif @scale > s.size
+      io << "0."
+      (@scale - s.size).times do
+        io << "0"
```
You could use a `Char` here.

akzhan replied:
No, string is better here because need no conversion from Char to String.

akzhan replied:
@Sija I'm wrong! Unbelievable! 

```
➜  crystal git:(bigfloat-frexp) ✗ bin/crystal 1.cr --no-debug --release
Using compiled compiler at `.build/crystal'
            char 370.13  (   2.7ms) (± 0.91%)       fastest
          string 150.32  (  6.65ms) (± 1.97%)  2.46× slower
  byte from char  12.61  ( 79.32ms) (± 2.65%) 29.36× slower
byte from string 133.63  (  7.48ms) (± 1.40%)  2.77× slower
            byte 132.72  (  7.53ms) (± 1.82%)  2.79× slower
```

```crystal
require "benchmark"

Benchmark.ips do |x|
  x.report("char") { io = IO::Memory.new; 1_000_000.times { io << '.' } }
  x.report("string") { io = IO::Memory.new; 1_000_000.times { io << "." } }
  x.report("byte from char") { io = IO::Memory.new; 1_000_000.times { io << '.'.bytes } }
  x.report("byte from string") { io = IO::Memory.new; 1_000_000.times { io << ".".byte_at(0) } }
  x.report("byte") { b = ".".byte_at(0); io = IO::Memory.new; 1_000_000.times { io << b } }
end
```

Sija replied:
@vegai Could you please include above tweaks?

vegai replied:
Huh, interesting :) Seems almost like a performance bug somewhere.

But will change these to chars then.

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,241 @@
+      io << s
+    else
+      offset = s.size - @scale
+      io << s[0...offset] << "." << s[offset..-1]
```
ditto

akzhan replied:
and again, String is better here (byte is more better).

vegai replied:
Done

akzhan commented:
LGTM

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+
+  # from `Float` is not supported due to precision loss risks. This call fails at compile time.
+  def initialize(num : Float)
+    {% raise "Initializing from Float is risky due to loss of precision -- use Int, String or BigRational" %}
```
I don't like this. We should support this. Why does round tripping through to_s loose precision? I thought float to_s was precise?

vegai replied:
Well... converting from a float literal should be safe every time, but floats can very easily be out of whack. 

```
a=0.1
b=0.0
puts a

b+=a
puts b
b+=a
puts b
b+=a
puts b
b+=a
puts b
```

=>

```
0.1
0.1
0.2
0.30000000000000004
0.4
```

Since perfect precision is BigDecimal's entire reason to exist, I'd rather just deny dangerous usage.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+  # Allows only valid number strings with an optional negative sign.
+  def initialize(str : String)
+    # disallow every non-valid string
+    if str !~ /^-?[0-9]+(\.[0-9]+)?$/
```
Actually here's a better way: 

```cr
reader = Char::Reader.new(str)
value = BigInteger.new(0)
scale = 0_u64 

if  reader.current_char == '-' 
  negative = true 
  reader.next_char
end 

has_decimal = false
until (char = reader.current_char) == '\0' 
  if char == '.' 
    raise ArgumentError.new if has_decimal 
    has_decimal = true 
    next 
  end
    
  int = char.to_i?
  raise ArgumentError.new unless int 

  value += int 
  scale += 1
  reader.next_char
end 

value = -value if negative 
```

I wrote this on my phone while on holiday so please fix all the mistakes!

vegai replied:
Just two errors, BigInteger => BigInt and missing reader.next_char in the until/if block :) 👍 

I benchmarked this against the regexp initializer, and it seemed to be a tiny bit faster with slower numbers and quite a lot slower with big numbers than the regexp-based initializer. 

```
                                        initialize 123.456 regexp   1.73  (578.12ms) (± 3.75%)  2.52× slower
                                        initialize 123.456 reader   1.95  (512.82ms) (± 3.77%)  2.24× slower
                                            initialize 123 regexp   3.82  (261.56ms) (± 2.27%)  1.14× slower
                                            initialize 123 reader   4.36  (229.35ms) (± 3.30%)       fastest
 initialize 0.12389127391827094872903841729038471290347808 regexp   1.37  (731.63ms) (± 1.15%)  3.19× slower
 initialize 0.12389127391827094872903841729038471290347808 reader   0.27  (  3.71s ) (± 0.05%) 16.20× slower
initialize 12387192378912738917128930.11729038471290347808 regexp   1.35  (739.18ms) (± 0.85%)  3.22× slower
initialize 12387192378912738917128930.11729038471290347808 reader   0.26  (  3.85s ) (± 0.38%) 16.80× slower
```

Do you still think this is a good idea?

RX14 replied:
Well at first glance I would think those benchmark results are highly suspect. I'll benchmark it myself and see what happened.

RX14 replied:
OK, I benchmarked it yourself, and you're right: it is slower. I suspect that the problem is the bigdecimal operations of `* 10` and `+ int` are slowing it down. Whereas with your solution, it's fast because it passes only the string to libgmp, which uses a fast path to parse it and turn it into a bigdecimal.

So I decided to take the best parts of both (my solution doesn't use regex, your solution uses libgmp as little as possible) and write a third solution, [here](https://github.com/RX14/crystal/commit/f6ee274674f7ca7b1bec824a4b08a0dffa61bf0e).

Benchmark code [here](https://gist.github.com/RX14/1429c54b7ab2f97578cbdf090f46966b):
```cr
Testing: 123.456
new2   4.23M (236.22ns) (± 4.98%)       fastest
 new   1.08M (929.53ns) (± 7.19%)  3.94× slower
 old   1.64M (608.84ns) (± 5.58%)  2.58× slower
Testing: 123
new2   8.81M (113.48ns) (± 7.95%)       fastest
 new   2.46M (405.82ns) (± 4.22%)  3.58× slower
 old   3.63M ( 275.8ns) (± 4.67%)  2.43× slower
Testing: 0.12389127391827094872903841729038471290347808
new2   2.56M (390.89ns) (± 2.44%)       fastest
 new 142.17k (  7.03µs) (± 4.53%) 17.99× slower
 old   1.22M (816.99ns) (± 7.29%)  2.09× slower
Testing: 12387192378912738917128930.11729038471290347808
new2   2.57M (388.77ns) (± 6.05%)       fastest
 new 134.51k (  7.43µs) (± 4.12%) 19.12× slower
 old   1.25M (799.48ns) (± 5.61%)  2.06× slower
```

My patch does change the semantics a bit: `.2` and `2.` syntaxes now work and yield `0.2` and `2` correctly. This can easilly be changed (it's an extra ` || index == 0 || (index == 1 && str[0] == '-'`), but the code is actually easier and simpler being a bit more permissive, and I think these two syntaxes are quite well-used anyway.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+    end
+  end
+
+  # from `Int`
```
All these short docstrings should be improved.

vegai replied:
Making all these full sentences.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+    check_division_by_zero other
+
+    scale = @scale - other.scale
+    n, d = @value, other.@value
```
Please don't use single character variable names.

vegai replied:
Fixed.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+    scale = @scale - other.scale
+    n, d = @value, other.@value
+
+    quotient, remainder = n.tdiv(d), n.remainder(d)
```
Could you use divmod here?

vegai replied:
Oh, certainly.

vegai replied:
Huh, unearthed a bug here. -1 / -2 gives us -0.5 currently.

vegai replied:
Fixed bugs and moved to divmod.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+  end
+
+  # from `Int`
+  def initialize(num : Int)
```
There should be a far faster way to initialize from integers than via to_s.

vegai replied:
Oh, whoops. Somebody was lazy :P Will fix.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+    remainder = remainder * TEN
+
+    i = 0
+    while remainder != ZERO && i < @@max_div_iterations
```
Shouldn't you raise if you reach max_div_iterations, not give the wrong result.

vegai replied:
It's not very exceptional, since  simple calculations like 1/3 hit this.

RX14 replied:
Oh never mind then. I didn't read the code properly.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+struct Float
+  # Casting from `Float` is not supported due to precision loss risks. This call fails at compile time.
+  def to_big_d
+    {% raise "Initializing from Float is risky due to loss of precision -- convert rather from Int or String" %}
```
Ditto

vegai replied:
Same as before, I guess?

vegai commented on spec/std/big/big_decimal_spec.cr:
```diff
@@ -0,0 +1,213 @@
+    (BigDecimal.new("-1") < BigDecimal.new("1")).should be_true
+  end
+
+  it "arithmetic that beats float precision" do
```
Yeah, I wasn't thinking at all there. Changed all the descriptions.

vegai commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+
+  include Comparable(BigDecimal)
+  include Comparable(Int)
+  include Comparable(String)
```
Replaced Int and String with Comparable(Numeric)

RX14 replied:
Oops, misremembered the name. Thanks!

vegai commented on spec/std/big/big_decimal_spec.cr:
```diff
@@ -0,0 +1,213 @@
+
+    (BigDecimal.new(1) > BigDecimal.new(1)).should be_false
+    (BigDecimal.new("1.00000000000000000000000000000000000001") > BigDecimal.new(1)).should be_true
+    (BigDecimal.new("0.99999999999999999999999999999999999999") > BigDecimal.new(1)).should be_false
```
Added < tests

Sija replied:
```cr
(BigDecimal.new("0.99999999999999999999999999999999999999") < BigDecimal.new(1)).should be_true
```

Sija commented on spec/std/big/big_decimal_spec.cr:
```diff
@@ -0,0 +1,223 @@
+  end
+
+  it "keeps precision" do
+    oneThousandth = BigDecimal.new("0.001")
```
please, use `snake_case` for variable names; we ain't in JavaScript Land Dorothy ;)

vegai replied:
Heh. Fixed.

vegai commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+struct BigDecimal
+  ZERO = BigInt.new(0)
+  TEN  = BigInt.new(10)
+  @@max_div_iterations = 100_u64
```
Yeah, I thought it was a bit awkward too. The original algorithm in Rust just had this value hardcoded inside the div method. I'll take a look at configuring it per div call. I cannot make this an optional parameter of `/` though, can I?

vegai replied:
Made it configurable per div call as suggested: added a div function with the optional parameter, and made / call that one.

vegai commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,249 @@
+
+  private property value : BigInt
+  private property scale : UInt64
+  getter value, scale
```
Hmm, I didn't quite get the gist here. Doesn't this combo make the value immutable from the outside world? Or does Crystal have better immutability protections that I've missed?

Sija commented on src/big/big_decimal.cr:
```diff
@@ -114,22 +114,36 @@ struct BigDecimal
-    return BigDecimal.new(quotient, scale) if remainder == ZERO
+    # quotient, remainder = n.tdiv(d), n.remainder(d)
+    quotient, remainder = n.divmod(d)
+    puts "n #{n} d #{d} q #{quotient} r #{remainder}"
```
Debug leftovers.

vegai replied:
Yeah, intermediate commits to save work. Removed afterwards

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,253 @@
+struct BigDecimal
+  ZERO                       = BigInt.new(0)
+  TEN                        = BigInt.new(10)
+  DEFAULT_MAX_DIV_ITERATIONS = 100_u64
```
Personally I wouldn't bother making this a constant but it's not worth any effort making it not a constant.

So ignore this comment.

akzhan commented:
Looks like You forgot about

```crystal
struct BigDecimal
  # Returns *num*. Useful for generic code that does `T.new(...)` with `T`
  # being a `Number`.
  def self.new(num : BigDecimal)
    num
  end
```

Some info here: #2292.

vegai commented:
@akzhan Thanks, added that.

mverzilli commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,262 @@
+
+struct BigDecimal
+  ZERO                       = BigInt.new(0)
+  TEN                        = BigInt.new(10)
```
Out of curiosity: Why a distinguished constant for 10? Is it a convenience to later express scientific notation?

Edit: nevermind, I should have continued reading to see how you used it :).

akzhan replied:
This simply to not instantiate TEN every time it is used (it is not primitive type, so its instantiation have a price). See https://github.com/crystal-lang/crystal/pull/4876#discussion_r135635824.

akzhan commented:
@vegai Can you apply @RX14 proposed parser and remove duplicated `it` as @mverzilli proposed?

mverzilli commented:
Yeah, as soon as you apply that we're ready to merge

vegai commented:
Ah, yes. I'll do that today, i.e. during the next 24h.

vegai commented:
Changes made & rebased to single commit.

mverzilli commented:
Thanks @vegai ! Couple of things:

1. I still don't see a test for `<=` (https://github.com/crystal-lang/crystal/pull/4876#discussion_r137932485), is that on purpose or maybe you missed that comment?

2. Next time, would you mind not squashing while we're at the middle of a review? For big PRs like this one it forces reviewers to re-read the whole diff instead of just focusing on the parts that change. We can always squash at the moment of merging if it makes sense.

akzhan commented:
Looks like CircleCI build failed due to LLVM 5.0, your branch based on too old master (that doesn't support LLVM 5.0 some days ago).

vegai commented:
@mverzilli Yes, I indeed missed that. Will add. Sorry about silly rebasing, will contain myself a bit better in the future.

@akzhan rebased from upstream's master.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,287 @@
+  end
+
+  def <=>(other : Int)
+    @value <=> other
```
I'm confused, isn't this incorrect? Wouldn't `@value` for `6.1234` be `61234`, and so if you compare it to `7_u32` you're comparing 61234 with 7, so it'll say `6.1234 > 7` ? I don't see any specs which do `BigDecimal.new(6.5) < 7` for example, they're all comparing bigdecimal ints with ints.

vegai replied:
Oh, indeed it is. I added that unit test and corrected the implementation as `self <=> BigDecimal.new(other)`

mverzilli commented:
@RX14 you had already approved but then requested some changes. This LGTM now, so feel free to merge once it's good for you.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,287 @@
+
+  # Creating a `BigDecimal` from `Float` is not supported due to precision loss risks. This call fails at compile time.
+  def initialize(num : Float)
+    {% raise "Initializing from Float is risky due to loss of precision -- use Int, String or BigRational" %}
```
The only thing i'm still concerned about is this ctor. I don't like the idea of refusing to convert just because it's lossy. There are plenty of lossy transforms already in the stdlib, they should be documented, not unimplemented.

mverzilli replied:
I like the strictness here. However, initializing from a BigRational is lossy as well, isn't it? (Think 1/3, for example).

akzhan replied:
I prefer idea of documenting losses and may be warnings about.

vegai replied:
I'm also drifting towards supporting these but warning/documenting. Would a compile-time warning be ok?

vegai replied:
@mverzilli true... and also, we don't even support initializations from BigRational :)

vegai replied:
I made a change where initialization and to_big_d from Float works, but outputs a warning to STDERR

asterite replied:
Please, no warnings. I can't imagine a program running with this code while STDERR gets flooded. I can also imagine the specs now output to STDERR?

vegai replied:
Ok, removed the warning. Devs will bear the responsiblity, then.

vegai replied:
@RX14 Does this satisfy your worries?

luislavena commented on src/big/big_decimal.cr:
```diff
@@ -73,10 +73,10 @@ struct BigDecimal
   end
 
-  # Creating a `BigDecimal` from `Float` is not supported due to precision loss risks. This call fails at compile time.
+  # Creating a `BigDecimal` from `Float` is not supported due to precision loss risks.
```
I would suggest change ctor description here. You **can** create a `BigDecimal` from `Float`, but a warning will be printed.

re: warning, we need a spec that checks for this scenario.

Sija commented on src/big/big_decimal.cr:
```diff
@@ -282,6 +282,6 @@ end
 struct Float
   # Casting from `Float` is not supported due to precision loss risks. This call fails at compile time.
```
>[...] This call fails at compile time.

no longer true.

vegai replied:
Fixed this comment and also a similar  one earlier in the file.

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,286 @@
+end
+
+struct Float
+  # Casting from `Float` is not recommended due to precision loss risks.
```
I'd suggest adding separate `NOTE` paragraph about precision loss.

vegai replied:
Modified the comment here and above in `initialize` a bit.

akzhan commented:
But where is `def hash(hasher)`?

vegai commented:
@akzhan Ah, right. I'm on it.

oprypin commented:
To clarify:
`hash` is currently not implemented. Although there is the default `Struct` Hash implementation, it just combines hashes of member variables.
You really want to ensure this invariant for your type: `if a==b then a.hash==b.hash`
Looking at the equality method, that is NOT currently ensured because it is possible for two bigdecimals with different members in to be equal.

akzhan commented:
@oprypin it's yet another "Number Normalization" issue was proposed by me and @funny-falcon a time ago.

Looks like I push it until 5 January :)

Sija commented on src/big/big_decimal.cr:
```diff
@@ -257,6 +257,20 @@ struct BigDecimal
 
+  # Factors out any extra powers of ten in the internal representation.
+  # For instance, value=100 scale=2 => value=1 scale=0
+  def factor_powers_of_ten
```
Shouldn't this be `private` and called only from inside `BigDecimal`?

vegai replied:
Ow, that was fast reactions :) I guess so. That makes it impossible to unit test it, though, doesn't it?

vegai commented:
Added some more tests. Changed to_s slightly so that it factors powers of ten away before outputting (so "1.00" will output 1). Made hasher use to_s and wrote some tests for that too -- one test specially trying out the a==b <=> h[a]==h[b] case.

oprypin commented on spec/std/big/big_decimal_spec.cr:
```diff
@@ -0,0 +1,307 @@
+  end
+
+  it "upkeeps hashing contract" do
+    # a == b <=> h[a] == h[b]
```
This test does actually check for `a = b ⇔ h[a] = h[b]` but that condition is impossible to satisfy because of pigeonhole principle.
The invariant is `a = b ⇒ h[a] = h[b]`. Just check `BigDecimal.new(1, 2).hash == BigDecimal.new(100, 4).hash` and be done with it.

vegai replied:
Spec changed as you recommended.

oprypin commented on spec/std/big/big_decimal_spec.cr:
```diff
@@ -0,0 +1,307 @@
+    hash[BigDecimal.new("0.12345")].should eq "bd2"
+    hash[BigDecimal.new("1.23456")].should eq "bd3"
+    hash[BigDecimal.new("123456")].should eq "bd4"
+    hash[BigDecimal.new("0")].should eq "bd5"
```
Can you explain what this test actually checks? The hashing method of `BigDecimal` could literally always return 0 and the test would still pass. Seems like more of a `Hash` test than a `#hash` test.

What I think this checks:
- That `BigDecimal.hash` does not raise an exception.
- That `BigDecimal#hash` outputs the same value every time (not randomized).
- That `BigDecimal#==` works correctly.

But there are more direct ways to check that.

vegai replied:
Hmm, seems to me like this is a useful test, even if indeed a bit redundant. It's checking that basic hashing scenarios with BigDecimal keys is working.

RX14 commented:
Does this PR not have any `to_i`/`to_u`/`to_f` impls? I think that "copping out" by going via string for any conversions in this PR is wrong.

RX14 commented:
Java's [BigDecimal](https://github.com/dmlloyd/openjdk/blob/7d7fbd09fcfd7f8cd02bf76ce10433ceeb33b3cf/jdk/src/java.base/share/classes/java/math/BigDecimal.java) impl is very interesting. It too goes via `String` for float and double. For int and long it uses a method called `setScale` to set the scale to 0, and then simply converts the `BigInt` to an integer.

vegai commented:
Added to_f/u/i implementations and a few specs. to_f goes via to_s while to_u and _i scale up a bit like Java's implementation.

RX14 commented on src/big/big_decimal.cr:
```diff
@@ -239,16 +245,23 @@ struct BigDecimal
+      (@value / TEN ** @scale)
+    else
+      -(@value.abs / TEN ** @scale)
+    end.to_i
```
No, we never chain a method call after `end`. The correct way to write this is to stick the `to_i` on the end of the if branches, which works because of explicit return.

If this wasn't the final (only) statement in the method, we simply assign to a variable inside if:
```cr
if @value >= 0
  int_value = (@value / TEN ** @scale).to_i
else
  int_value = -(@value.abs / TEN ** @scale).to_i
end

# use int_value
```

vegai replied:
Righto. Fixed.

vegai commented:
Should this PR be in the "Numbers" project?

Sija commented on src/big/big_decimal.cr:
````diff
@@ -0,0 +1,328 @@
+  # defines a maximum number of iterations in case the division is not exact.
+  #
+  # ```
+  # BigDecimal(1).div(BigDecimal(2)) => BigDecimal(@value=5, @scale=2)
````
Missing `#` before `=>`. On the line below as well.

vegai replied:
Roger, fixed.

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,328 @@
+    div other
+  end
+
+  # Divides self with another `BigDecimal`, with a optionally configurable max_div_iterations, which
```
`max_div_iterations` -> `* max_div_iterations*`

vegai replied:
Fixed.

Sija commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,328 @@
+    hasher.string(self.to_s)
+  end
+
+  # Returns the quotient as absolutely negative if self and other have different signs,
```
`quotient` -> `*quotient*`

vegai replied:
Fixed (also line below).

Thanks!

onemanstartup commented:
Thank you! 🎉

mverzilli commented:
Thank you @vegai! Amazing work, and outstanding patience to bear with reviews :)

akzhan commented on src/big/big_decimal.cr:
```diff
@@ -0,0 +1,328 @@
+  end
+
+  def hash(hasher)
+    hasher.string(self.to_s)
```
Of course it must be rewritten using number normalization.

But it will be anyway later.

Sija commented:
This PR added `src/big_decimal.cr` which got obsolete because of #5121.

Sija commented:
btw, why `BigDecimal` doesn't inherit from `Number`?

akzhan commented:
@oprypin Done - #5276
